litestar/utils/predicates.py

```diff
@@ -124,6 +124,9 @@ def is_dataclass_class(annotation: Any) -> TypeGuard[type[DataclassProtocol]]:
         ``True`` if instance or type of ``dataclass``.
     """
     try:
+        origin = get_origin_or_inner_type(annotation)
```

Comment:
Why are we using `get_origin_or_inner_type()` here, is it just to remove `Annotated` or other wrapper types?

Can we not just use `utils.typing.unwrap_annotation` for `annotation, _, _ = unwrap_annotation(annotation)`?

Reply:
This was to handle cases like `GenericDataclass[int]`. Before it returned `False`, but now I think about it, it could be fine to get the caller to this function to ensure that it's the origin that is being passed. I'm fine with either approach tbh.

Reply:
Oh ok, got you, so we actually do want some flavor of `get_origin()`.

Reply:
Yeah, this same issue was there for `TypedDict` as well.

Reply:
What do you think about splitting `unwrap_and_get_instantiable_origin_if_it_exists_else_the_origin` into two different functions: `unwrap_and_get_origin` and `unwrap_and_get_instantiable_origin`? In both cases `None` would be returned if there's not origin or instantiable origin, respectively.

Reply:
So I guess the question is, should we consider a `_GenericAlias` that represents a dataclass type, as a dataclass type?

Reply:
> 

What do you think about splitting unwrap_and_get_instantiable_origin_if_it_exists_else_the_origin into two different functions: unwrap_and_get_origin and unwrap_and_get_instantiable_origin? In both cases None would be returned if there's not origin or instantiable origin, respectively.

The thing is, this work has already been done on `FieldDefinition`, which is what we are using to call these methods anyway:

```py
# _openapi.schema_generation.schema
...
        elif is_dataclass_class(field_definition.annotation):
            result = self.for_dataclass(field_definition.annotation)
```

Here `field_definition` is an instance of `typing.FieldDefinition` and we have already calculated `origin`, it is available at `field_defintion.origin`. Also, `field_definition.annotation` is by definition already unwrapped (`field_definition.raw` being the original untouched annotation).

So, I'm starting to think we should be using that. Even maybe adding some extra predicates to that type, e.g., `FieldDefinition.is_dataclass_type()`, which could be:

```py
class FieldDefinition:
    def is_dataclass_type(self) -> bool:
        return is_dataclass_class(self.origin or self.annotation)
```

Reply:
Oh, I hadn't looked too much into the `FieldDefinition`. I think adding the `is_dataclass_type` and the `is_typeddict_type` might be nice. The other predicates like `is_struct_class`, `is_pydantic_class` etc. all work since it does a `is_class_and_subclass` which does the unwrapping if needed.

Reply:
Unfortunately, we will still be doing these checks again in the `for_*` methods in the `get_unwrapped_annotation_and_type_hints` helper function. To avoid this we could do the following:

- add a `get_type_hints` method to the `FieldDefinition` which will call `get_type_hints` or `get_type_hints_with_generics_resolved` since it already knows whether the annotation is generic or not.
- pass in the `FieldDefinition` instance to the `for_*` functions which is what we do for the plugins anyway

Reply:
That sounds reasonable. Given a type, `FieldDefinition(typ).get_type_hints()` would basically be the same as calling `typing_extensions.get_type_hints(typ)`?

You could make the generic resolution optional via flag?

```py
class FieldDefinition:
    def get_type_hints(self, include_extras: bool = False, resolve_generics: bool = False) -> dict[str, Any]: ...
```

Reply:
Okay so I was thinking of the following changes:
- remove the `get_origin_or_inner_type` from the `is_dataclass_class` and `is_typed_dict` and document in the docstrings that if the type is generic, then the origin is what has to be given
- add `is_dataclass_type` and `is_typeddict_type` helpers to the `FieldDefinition`
- add `get_type_hints` with the above signature you've mentioned
- pass the `FieldDefinition` to the `for_*` methods on the schema generation which means the `get_unwrapped_annotation_and_type_hints` can be removed

Sounds good?

Reply:
> remove the get_origin_or_inner_type from the is_dataclass_class and is_typed_dict and document in the docstrings that if the type is generic, then the origin is what has to be given

How about we leave them alone and deprecate them in favor of `FieldDefinition.{is_dataclass_type,is_typeddict_type}` in-case anyone is using them, and re-implement the logic in `FieldDefinition` methods. They are pretty light-on on functionality.

The rest SGTM!

Reply:
If we deprecate them, what about cases where we or the user needs to check if a type is a dataclass/typeddict class, but it's not wrapped in a `FieldDefinition` instance?

Reply:
My opinion is that we should strive to have one way to interact with types internally, via `FieldDefinition`. Much of the untidiness we've seen in this area of the code-base (pre-existing, not yours) stems from the fact that we have both the OO approach to reasoning about our types (`FieldDefinition`), and the functional approach via `utils.typing`/`utils.predicates` etc. And we've seen in our discussion here how that is leading to things like duplication of functionality and making code hard to reason about.

While the functions I propose to deprecate aren't in our API docs, they also aren't in any sort of private namespace - which is why I don't think we should change their behavior. And, if we aren't using functionality ourselves internally (the only place we use them is for this openapi logic), I don't think there is merit in carrying them forward ad infinitum, hence my suggestion to deprecate in favor of the `FieldDefinition` properties.

We'll still have a public interface through which users can leverage our type introspection logic, which would be via `FieldDefinition` and users can easily implement the functions themselves if they don't want to go that route.

Reply:
Yeah, that sounds good to me. I think it will be good if we start enforcing the use of `FieldDefinition` in our internal API.